src/mender-update/standalone/states.cpp

```diff
@@ -0,0 +1,555 @@
+			UpdateResult(ctx.result_and_error, {Result::Failed | Result::CleanupFailed, err});
+			final_event = StateEvent::Failure;
+			data.failed = true;
+			// Fall through so that we update the DB.
+		}
+	}
+
+	error::Error err;
+	if (data.rolled_back) {
+		// Successful rollback.
+		auto &db = ctx.main_context.GetMenderStoreDB();
+		err = db.Remove(context::MenderContext::standalone_state_key);
+	} else {
+		if (data.failed) {
+			// Unsuccessful rollback or missing rollback support.
+			data.artifact_name += ctx.main_context.broken_artifact_name_suffix;
+			if (data.artifact_provides) {
+				data.artifact_provides.value()["artifact_name"] = data.artifact_name;
+			}
+			// Fall through to success case.
+		}
+		// Commit artifact data and remove state data
+		err = ctx.main_context.CommitArtifactData(
+			data.artifact_name,
+			data.artifact_group,
+			data.artifact_provides,
+			data.artifact_clears_provides,
+			[](database::Transaction &txn) {
+				return txn.Remove(context::MenderContext::standalone_state_key);
+			});
+	}
+	if (err != error::NoError) {
+		err = err.WithContext("Error while updating database");
+		UpdateResult(ctx.result_and_error, {Result::Failed | Result::RollbackFailed, err});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	}
+
+	UpdateResult(ctx.result_and_error, {Result::Cleaned, error::NoError});
+	poster.PostEvent(final_event);
+}
+
+void ScriptRunnerState::OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) {
+	auto err = ctx.script_runner->RunScripts(state_, action_, on_error_);
+	if (err != error::NoError) {
+		log::Error("Error executing script: " + err.String());
+		UpdateResult(ctx.result_and_error, {result_on_error_, err});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	}
+
+	poster.PostEvent(StateEvent::Success);
+}
+
+void ExitState::OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) {
+	auto err =
+		ctx.main_context.GetMenderStoreDB().WriteTransaction([&ctx](database::Transaction &txn) {
+			auto exp_bytes = txn.Read(context::MenderContext::standalone_state_key);
+			if (!exp_bytes) {
+				if (exp_bytes.error().code == database::MakeError(database::KeyError, "").code) {
+					// If the stata data is not saved, just do nothing here.
+					return error::NoError;
+				} else {
+					return exp_bytes.error();
+				}
+			}
+
+			// If there is state data, resave it with `failed` set to false. The rationale
+			// behind this is that if we have already recorded failure for this run, it will be
+			// returned in the error code. That does not mean that we should record error for
+			// the next run, which is independent. An example is rollback, if we are somewhere
+			// in the rollback flow, we are likely to have a failure here, because the *install*
+			// failed. But when we now exit, and then later resume the rollback, the rollback
+			// should return success, not failure.
+			ctx.state_data.failed = false;
+			return SaveStateData(txn, ctx.state_data);
```

Comment:
We could guard this to spare some writing
```suggestion
			if (ctx.state_data.failed) {
				ctx.state_data.failed = false;
				return SaveStateData(txn, ctx.state_data);
			}
```